subiquity/models/mirror.py

```diff
@@ -49,51 +152,88 @@ class MirrorModel(object):
-        self.primary: List[Any] = [
-            {
-                "arches": PRIMARY_ARCHES,
-                "uri": "http://archive.ubuntu.com/ubuntu",
-            }, {
-                "arches": ["default"],
-                "uri": "http://ports.ubuntu.com/ubuntu-ports",
-            },
-        ]
+        self.primary_elected: Optional[PrimaryElement] = None
+        self.primary_candidates: List[PrimaryElement] = \
+            self._default_primary_entries()
+
+        self.primary_staged: Optional[PrimaryElement] = None
 
         self.architecture = get_architecture()
-        self.default_mirror = self.get_mirror()
+        # Only useful for legacy primary sections
+        self.default_mirror = \
+            LegacyPrimarySection.new_from_default(parent=self).uri
+
+    def _default_primary_entries(self) -> List[PrimaryEntry]:
+        return [
+            PrimaryEntry(parent=self, country_mirror=True),
+            PrimaryEntry(uri=DEFAULT_SUPPORTED_ARCHES_URI,
+                         arches=PRIMARY_ARCHES, parent=self),
+            PrimaryEntry(uri=DEFAULT_PORTS_ARCHES_URI,
+                         arches=PORTS_ARCHES, parent=self),
+        ]
+
+    def get_default_primary_candidates(
+            self, legacy: Optional[bool] = None) -> Sequence[PrimaryElement]:
+        want_legacy = legacy if legacy is not None else self.legacy_primary
+        if want_legacy:
+            return [LegacyPrimarySection.new_from_default(parent=self)]
+        else:
+            return self._default_primary_entries()
 
     def load_autoinstall_data(self, data):
+        self.legacy_primary = data.pop("version", 1) < 2
         if "disable_components" in data:
             self.disabled_components = set(data.pop("disable_components"))
         if "primary" in data:
-            self.primary = data.pop("primary")
+            if self.legacy_primary:
+                # Legacy sections only support a single candidate
+                self.primary_candidates = \
+                    [LegacyPrimarySection(data.pop("primary"), parent=self)]
+            else:
+                self.primary_candidates = []
+                for section in data.pop("primary"):
+                    entry = PrimaryEntry.from_config(section, parent=self)
+                    self.primary_candidates.append(entry)
+        else:
+            self.primary_candidates = self.get_default_primary_candidates()
+
         merge_config(self.config, data)
 
-    def get_apt_config(self):
+    def _get_apt_config_common(self) -> Dict[str, Any]:
+        assert "disable_components" not in self.config
+        assert "primary" not in self.config
+
         config = copy.deepcopy(self.config)
-        config["primary"] = copy.deepcopy(self.primary)
         config["disable_components"] = sorted(self.disabled_components)
         return config
 
-    def mirror_is_default(self):
-        return self.get_mirror() == self.default_mirror
+    def get_apt_config_staged(self) -> Dict[str, Any]:
```

Comment:
I added it, thanks. 
I took the opportunity to replace "elected" by "final" in apt code since "elected" is more of a mirror thing, that does not make sense in other contexts.